currencies/rate_converter_test.go

```diff
@@ -13,354 +13,273 @@ import (
-			calledURLs = append(calledURLs, req.RequestURI)
-			rw.WriteHeader(http.StatusOK)
-			rw.Write([]byte(
-				`{
-					"dataAsOf":"2018-09-12",
-					"conversions":{
-						"USD":{
-							"GBP":0.77208
-						},
-						"GBP":{
-							"USD":1.2952
-						},
-						"badJsonHere"
-					}
-				}`,
-			))
-		}),
-	)
-
-	defer mockedHttpServer.Close()
+	mockServerHandler := func(mockResponse []byte, code int) http.Handler {
+		return http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+			rw.WriteHeader(code)
+			rw.Write(mockResponse)
+		})
+	}
 
-	// Execute:
-	currencyConverter := currencies.NewRateConverter(
-		&http.Client{},
-		mockedHttpServer.URL,
-		time.Duration(24)*time.Hour,
-	)
+	testCases := []struct {
+		mockResponse     []byte
+		mockResponseCode int
+		updateInterval   int // In seconds
+		fetchURL         string
+		expectedTicks    int
+		description      string
+	}{
+		{
+			mockResponse:     getMockRates(),
+			mockResponseCode: 200,
+			updateInterval:   1,
+			expectedTicks:    1,
+			description:      "Fetching currency rates successfully",
+		},
+		{
+			mockResponse:     []byte(""),
+			mockResponseCode: 200,
+			updateInterval:   100,
+			description:      "Currency rates endpoint returns empty response",
+		},
+		{
+			mockResponse:     []byte("`invalid-json:"),
+			mockResponseCode: 200,
+			updateInterval:   100,
+			description:      "Currency rates endpoint returns invalid json",
+		},
+		{
+			mockResponse:     nil,
+			mockResponseCode: 200,
+			updateInterval:   100,
+			description:      "Currency rates endpoint returns nil response",
+		},
+		{
+			mockResponseCode: 404,
+			updateInterval:   4,
+			description:      "Currency rates endpoint returns 404",
+		},
+		{
+			mockResponseCode: 400,
+			updateInterval:   1,
+			description:      "Currency rates endpoint returns 400",
+		},
+		{
+			mockResponse:     getMockRates(),
+			mockResponseCode: 200,
+			description:      "Fetch interval set to 0",
+		},
+		{
+			fetchURL:    "invalid-url",
+			description: "Invalid currency rates endpoint",
+		},
+	}
 
-	// Verify:
-	assert.Equal(t, 1, len(calledURLs), "sync URL should have been called %d times but was %d", 1, len(calledURLs))
-	assert.Equal(t, currencyConverter.LastUpdated(), (time.Time{}), "LastUpdated() shouldn't return a time set")
-	assert.Nil(t, currencyConverter.Rates(), "Rates() should return nil")
-}
+	for _, test := range testCases {
+		mockedHttpServer := httptest.NewServer(mockServerHandler(test.mockResponse, test.mockResponseCode))
 
-func TestFetch_InvalidRemoteResponseContent(t *testing.T) {
+		if test.fetchURL == "" {
+			test.fetchURL = mockedHttpServer.URL
+		}
+		// Execute:
+		beforeExecution := time.Now()
+		currencyConverter := currencies.NewRateConverter(
+			&http.Client{},
+			test.fetchURL,
+			time.Duration(test.updateInterval)*time.Second,
+		)
+
+		// Verify:
+		if test.expectedTicks == 0 {
+			// Check for the next 200 milliseconds that no fetch was triggered
```

Comment:
Why 200ms?

Be careful here, time based unit tests are notoriously fragile. I understand the code was hard to instrument for testing. Ideally we would factor out the scheduling mechanism for the unit tests. Let's see if we can find a more resilient trigger.. but if not, I know you wanted to refactor this code in the future to fix the testability issues.